src/main/java/com/thoughtworks/gauge/scan/CustomClassInitializerScanner.java

```diff
@@ -0,0 +1,39 @@
+package com.thoughtworks.gauge.scan;
+
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.reflections.Reflections;
+
+import com.thoughtworks.gauge.ClassInitializer;
+import com.thoughtworks.gauge.CustomClassInitializer;
+import com.thoughtworks.gauge.registry.ClassInitializerRegistry;
+
+public class CustomClassInitializerScanner implements IScanner {
+    @Override
+    public void scan(Reflections reflections) {
+        scanForInitializer(reflections);
+    }
+
+    private void scanForInitializer(Reflections reflections) {
+        Set<Class<?>> initializers = reflections.getTypesAnnotatedWith(CustomClassInitializer.class);
+
+        if (initializers.size() > 1) {
+            throw new ScanningException("Only 1 class can be annotated with @CustomClassInitializer");
+        }
+
```

Comment:
Hello @kashishm,
The highlighted snippet will not check for null, it will check if there are more than 1 class annotated with @CustomClassInitializer, in this case the user should be notified with an exception showing that only 1 CustomInitializer should be annotated.

For the case that you are pointing at in your comment, we are already setting the classInitializer to null in the registry which is set in case no custom intitializer is set.
But we can improve such logic and _remove the null checks in ClassInstanceManager (line 4 to 8) as shown below_:

**Current ClassInstanceManager Class:**
```
1.     public Object get(Class<?> declaringClass) throws Exception {
2.         Object classInstance = classInstanceMap.get(declaringClass);
3.         if (classInstance == null) {
4.             if (getInitializer() != null) {
5.                 classInstance = getInitializer().initialize(declaringClass);
6.             } else {
7.                 classInstance = Class.forName(declaringClass.getName()).newInstance();
8.             }
9.             classInstanceMap.put(declaringClass, classInstance);
10.         }
11.         return classInstance;
12.     }
```
**Proposed ClassInstanceManager Class:**
```
1.     public Object get(Class<?> declaringClass) throws Exception {
2.         Object classInstance = classInstanceMap.get(declaringClass);
3.         if (classInstance == null) {          
4.             classInstance = getInitializer().initialize(declaringClass);
5.             classInstanceMap.put(declaringClass, classInstance);
6.         }
7.         return classInstance;
8.     }

```
To apply the proposed logic, the following should be done:
- Create a new class _DefaultClassInitializer_:
```
public class DefaultClassInitializer implements ClassInitializer {

    @Override
    public Object initialize(Class<?> classToInitialize) throws Exception {
        return Class.forName(classToInitialize.getName()).newInstance();
    }
}
```

- Modify CustomClassInitializerScanner, and include the following code in the catch close (NoSuchElementException  Line 20 below)

```
1.  private void scanForInitializer(Reflections reflections) {
2.         Set<Class<?>> initializers = reflections.getTypesAnnotatedWith(CustomClassInitializer.class);
3. 
4.         if (initializers.size() > 1) {
5.             throw new ScanningException("Only 1 class can be annotated with @CustomClassInitializer");
6.         }
7. 
8.         try {
9.             Class<?> initializer = initializers.iterator().next();
10.             if (!ClassInitializer.class.isAssignableFrom(initializer)) {
11.                 throw new ScanningException("Custom class initializer should implement ClassInitializer interface");
12.             }
13.             ClassInitializerRegistry.classInitializer((ClassInitializer) initializer.newInstance());
14.         } catch (InstantiationException e) {
15.             throw new ScanningException(e.getMessage());
16.         } catch (IllegalAccessException e) {
17.             throw new ScanningException(e.getMessage());
18.         } catch (NoSuchElementException e) {
19.             // Class initializer is not defined, use default!
20.             ClassInitializerRegistry.classInitializer(new DefaultClassInitializer());
21.         }
22.     }
```